From 97975f9ab954d1a955cc203a77dfad5d2750c55f Mon Sep 17 00:00:00 2001
From: Mark Reynolds <mreynolds@redhat.com>
Date: Tue, 25 Mar 2025 14:29:19 -0400
Subject: [PATCH] Issue 6500 - Fix covscan and ASAN issue

Description:

- Fix memory leak in alias entry plugin
- Fix lock issue in replication changelog ldif import
- Fix memory leak in replication parse_changes_string()
- Fix memory leak in views plugin
- Remove dead code in auditlog.c
- Fix memory leak in dbmdb_build_dbname()
- Fix uninitialized variable warning in log.c
- Fix several memory leaks in modify.c
- Fix memory leak with result text
- Fix memory leak in snmp main.c when loading config

Relates: https://github.com/389ds/389-ds-base/issues/6500

Reviewed by: progier(Thanks!)
---
 .../plugins/alias_entries/alias-entries.c     |  5 ++++-
 ldap/servers/plugins/memberof/memberof.c      |  1 +
 .../plugins/memberof/memberof_config.c        | 16 +++++++++++++-
 ldap/servers/plugins/replication/cl5_api.c    | 22 +++++++++++--------
 ldap/servers/plugins/replication/replutil.c   |  2 ++
 ldap/servers/plugins/views/views.c            |  1 +
 ldap/servers/slapd/auditlog.c                 |  9 ++------
 .../slapd/back-ldbm/db-mdb/mdb_instance.c     |  5 ++++-
 ldap/servers/slapd/back-ldbm/ldbm_config.c    |  3 ++-
 ldap/servers/slapd/log.c                      |  8 +++----
 ldap/servers/slapd/modify.c                   |  7 ++++++
 ldap/servers/slapd/operation.c                |  1 +
 ldap/servers/slapd/pblock.c                   |  1 +
 ldap/servers/slapd/plugin.c                   |  5 -----
 ldap/servers/slapd/slapi-memberof.c           |  4 +---
 ldap/servers/snmp/main.c                      |  6 ++---
 16 files changed, 61 insertions(+), 35 deletions(-)

diff --git a/ldap/servers/plugins/alias_entries/alias-entries.c b/ldap/servers/plugins/alias_entries/alias-entries.c
index ddd775df4..8b5862960 100644
--- a/ldap/servers/plugins/alias_entries/alias-entries.c
+++ b/ldap/servers/plugins/alias_entries/alias-entries.c
@@ -121,6 +121,7 @@ alias_entry_srch(Slapi_PBlock *pb)
             slapi_log_error(SLAPI_LOG_PLUGIN, PLUGINNAME,
                             "alias_entry_srch - %s\n", errorbuf);
 
+            slapi_sdn_free(&dn1);
             slapi_send_ldap_result(pb, rc, NULL, errorbuf, 0, NULL);
             slapi_pblock_set(pb, SLAPI_PLUGIN_OPRETURN, &rc);
             return SLAPI_PLUGIN_FAILURE;
@@ -128,7 +129,8 @@ alias_entry_srch(Slapi_PBlock *pb)
     } while (dn2 != NULL && i++ < MAXALIASCHAIN);
 
     if (dn1 == search_target) {
-        /* Source dn is not an alias */
+        /* Source dn is not an alias */\
+        slapi_sdn_free(&dn2);
         return 0;
     }
 
@@ -138,6 +140,7 @@ alias_entry_srch(Slapi_PBlock *pb)
         slapi_pblock_set(pb, SLAPI_SEARCH_TARGET_SDN, dn1);
     } else {
         /* Here we hit an alias chain longer than MAXALIASCHAIN */
+        slapi_sdn_free(&dn1);
         slapi_sdn_free(&dn2);
     }
 
diff --git a/ldap/servers/plugins/memberof/memberof.c b/ldap/servers/plugins/memberof/memberof.c
index 16dae2195..3775e52c9 100644
--- a/ldap/servers/plugins/memberof/memberof.c
+++ b/ldap/servers/plugins/memberof/memberof.c
@@ -931,6 +931,7 @@ perform_needed_fixup(void)
         }
         be = slapi_get_next_backend(cookie);
     }
+    slapi_ch_free_string(&cookie);
     slapi_ch_free_string(&td.bind_dn);
     slapi_ch_free_string(&td.filter_str);
     memberof_free_config(&config);
diff --git a/ldap/servers/plugins/memberof/memberof_config.c b/ldap/servers/plugins/memberof/memberof_config.c
index fca5251d0..b74c24158 100644
--- a/ldap/servers/plugins/memberof/memberof_config.c
+++ b/ldap/servers/plugins/memberof/memberof_config.c
@@ -576,10 +576,24 @@ memberof_apply_config(Slapi_PBlock *pb __attribute__((unused)),
 
             /* Add beginning of filter. */
             bytes_out = snprintf(filter_str, filter_str_len - bytes_out, "(|");
+            if (bytes_out<0) {
+                slapi_log_err(SLAPI_LOG_ERR, MEMBEROF_PLUGIN_SUBSYSTEM,
+                              "snprintf unexpectly failed in memberof_apply_config.\n");
+                *returncode = LDAP_UNWILLING_TO_PERFORM;
+                goto done;
+            }
 
             /* Add filter section for each groupattr. */
             for (i = 0; theConfig.groupattrs && theConfig.groupattrs[i]; i++) {
-                bytes_out += snprintf(filter_str + bytes_out, filter_str_len - bytes_out, "(%s=*)", theConfig.groupattrs[i]);
+                size_t bytes_read = snprintf(filter_str + bytes_out, filter_str_len - bytes_out, "(%s=*)",
+                                             theConfig.groupattrs[i]);
+                if (bytes_read<0) {
+                    slapi_log_err(SLAPI_LOG_ERR, MEMBEROF_PLUGIN_SUBSYSTEM,
+                                  "snprintf unexpectly failed in memberof_apply_config.\n");
+                    *returncode = LDAP_UNWILLING_TO_PERFORM;
+                    goto done;
+                }
+                bytes_out += bytes_read;
             }
 
             /* Add end of filter. */
diff --git a/ldap/servers/plugins/replication/cl5_api.c b/ldap/servers/plugins/replication/cl5_api.c
index 1b6413164..d88ce7d98 100644
--- a/ldap/servers/plugins/replication/cl5_api.c
+++ b/ldap/servers/plugins/replication/cl5_api.c
@@ -540,7 +540,18 @@ cl5ImportLDIF(const char *clDir, const char *ldifFile, Replica *replica)
         return CL5_BAD_DATA;
     }
 
+    /* Set changelog state to import */
+    pthread_mutex_lock(&(cldb->stLock));
+
+    if (cldb->dbState == CL5_STATE_IMPORT) {
+        pthread_mutex_unlock(&(cldb->stLock));
+        slapi_log_err(SLAPI_LOG_ERR, repl_plugin_name_cl,
+                      "cl5ImportLDIF - changelog import already in progress\n");
+        return CL5_IGNORE_OP;
+    }
+
     if (cldb->dbState != CL5_STATE_OPEN) {
+        pthread_mutex_unlock(&(cldb->stLock));
         slapi_log_err(SLAPI_LOG_ERR, repl_plugin_name_cl,
                       "cl5ImportLDIF - Changelog is not initialized\n");
         return CL5_BAD_STATE;
@@ -549,6 +560,7 @@ cl5ImportLDIF(const char *clDir, const char *ldifFile, Replica *replica)
     /* open LDIF file */
     file = ldif_open(ldifFile, "r");
     if (file == NULL) {
+        pthread_mutex_unlock(&(cldb->stLock));
         slapi_log_err(SLAPI_LOG_ERR, repl_plugin_name_cl,
                       "cl5ImportLDIF - Failed to open (%s) ldif file; system error - %d\n",
                       ldifFile, errno);
@@ -556,15 +568,6 @@ cl5ImportLDIF(const char *clDir, const char *ldifFile, Replica *replica)
         goto done;
     }
 
-    /* Set changelog state to import */
-    pthread_mutex_lock(&(cldb->stLock));
-
-    if (cldb->dbState == CL5_STATE_IMPORT) {
-        pthread_mutex_unlock(&(cldb->stLock));
-        slapi_log_err(SLAPI_LOG_ERR, repl_plugin_name_cl,
-                      "cl5ImportLDIF - changelog import already in progress\n");
-        return CL5_IGNORE_OP;
-    }
     cldb->dbState = CL5_STATE_IMPORT;
 
     pthread_mutex_unlock(&(cldb->stLock));
@@ -3723,6 +3726,7 @@ _cl5LDIF2Operation(char *ldifEntry, slapi_operation_parameters *op, char **replG
         }
     }
     slapi_ch_free_string(&ldifEntryWork);
+    slapi_ch_free_string(&rawDN);
     return rval;
 }
 
diff --git a/ldap/servers/plugins/replication/replutil.c b/ldap/servers/plugins/replication/replutil.c
index 3f40ba071..fe118c02d 100644
--- a/ldap/servers/plugins/replication/replutil.c
+++ b/ldap/servers/plugins/replication/replutil.c
@@ -342,6 +342,7 @@ parse_changes_string(char *str)
             if (strcasecmp(line, "-") == 0) {
                 if (slapi_mod_isvalid(&mod)) {
                     slapi_mods_add_smod(mods, &mod);
+                    slapi_mod_init(&mod, 0);
                 } else {
                     /* need to cleanup */
                     slapi_mod_done(&mod);
@@ -382,6 +383,7 @@ parse_changes_string(char *str)
             }
             line = ldif_getline(&next);
         }
+        slapi_mod_done(&mod);
     }
 
     return mods;
diff --git a/ldap/servers/plugins/views/views.c b/ldap/servers/plugins/views/views.c
index 007a9d6f5..5e4499ddb 100644
--- a/ldap/servers/plugins/views/views.c
+++ b/ldap/servers/plugins/views/views.c
@@ -1139,6 +1139,7 @@ views_dn_views_cb(Slapi_Entry *e, void *callback_data)
             if (attrType && !strcasecmp(attrType, VIEW_FILTER_ATTR)) {
                 if (!slapi_attr_get_bervals_copy(dnAttr, &dnVals)) {
                     /* add filter */
+                    slapi_ch_free_string(&pView->viewfilter);
                     pView->viewfilter = slapi_ch_strdup(dnVals[0]->bv_val);
                 }
 
diff --git a/ldap/servers/slapd/auditlog.c b/ldap/servers/slapd/auditlog.c
index 747448998..3a34959f6 100644
--- a/ldap/servers/slapd/auditlog.c
+++ b/ldap/servers/slapd/auditlog.c
@@ -643,13 +643,8 @@ write_audit_file_json(Slapi_PBlock *pb, Slapi_Entry *entry, int logtype,
         case SLAPI_OPERATION_DELETE:
             tmp = change;
             del_obj = json_object_new_object();
-            if (tmp && tmp[0]) {
-                json_object_object_add(del_obj, "dn", json_object_new_string(target_dn));
-                json_object_object_add(log_json, "delete", del_obj);
-            } else {
-                json_object_object_add(del_obj, "dn", json_object_new_string(target_dn));
-                json_object_object_add(log_json, "delete", del_obj);
-            }
+            json_object_object_add(del_obj, "dn", json_object_new_string(target_dn));
+            json_object_object_add(log_json, "delete", del_obj);
             break;
 
         case SLAPI_OPERATION_MODDN:
diff --git a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_instance.c b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_instance.c
index 0315693e0..682fd70e2 100644
--- a/ldap/servers/slapd/back-ldbm/db-mdb/mdb_instance.c
+++ b/ldap/servers/slapd/back-ldbm/db-mdb/mdb_instance.c
@@ -210,7 +210,10 @@ char *dbmdb_build_dbname(backend *be, const char *filename)
         res = slapi_ch_smprintf("%s/%s%s", inst->inst_name, filename, suffix);
     }
     pt = (char*)slapi_utf8StrToLower((unsigned char*)res);
-    slapi_ch_free_string(&res);
+    if (pt != res) {
+        slapi_ch_free_string(&res);
+    }
+
     return pt;
 }
 
diff --git a/ldap/servers/slapd/back-ldbm/ldbm_config.c b/ldap/servers/slapd/back-ldbm/ldbm_config.c
index d16241df1..d7d138286 100644
--- a/ldap/servers/slapd/back-ldbm/ldbm_config.c
+++ b/ldap/servers/slapd/back-ldbm/ldbm_config.c
@@ -1294,7 +1294,8 @@ ldbm_config_search_entry_callback(Slapi_PBlock *pb __attribute__((unused)),
             if (attrs) {
                 for (size_t i = 0; attrs[i]; i++) {
                     if (ldbm_config_moved_attr(attrs[i])) {
-                        slapi_pblock_set(pb, SLAPI_RESULT_TEXT, "at least one required attribute has been moved to the DB scecific configuration entry");
+                        char *msg = slapi_ch_strdup("at least one required attribute has been moved to the DB specific configuration entry");
+                        slapi_pblock_set(pb, SLAPI_RESULT_TEXT, msg);
                         break;
                     }
                 }
diff --git a/ldap/servers/slapd/log.c b/ldap/servers/slapd/log.c
index 2d4144d35..178d29b89 100644
--- a/ldap/servers/slapd/log.c
+++ b/ldap/servers/slapd/log.c
@@ -3246,7 +3246,7 @@ log__open_accesslogfile(int logfile_state, int locked)
     char tbuf[TBUFSIZE];
     struct logfileinfo *logp;
     char buffer[BUFSIZ];
-    int rc;
+    int rc = 0;
 
     if (!locked)
         LOG_ACCESS_LOCK_WRITE();
@@ -3410,7 +3410,7 @@ log__open_securitylogfile(int logfile_state, int locked)
     char tbuf[TBUFSIZE];
     struct logfileinfo *logp;
     char buffer[BUFSIZ];
-    int rc;
+    int rc = 0;
 
     if (!locked)
         LOG_SECURITY_LOCK_WRITE();
@@ -6317,7 +6317,7 @@ log__open_auditlogfile(int logfile_state, int locked)
     char tbuf[TBUFSIZE];
     struct logfileinfo *logp;
     char buffer[BUFSIZ];
-    int rc;
+    int rc = 0;
 
     if (!locked)
         LOG_AUDIT_LOCK_WRITE();
@@ -6476,7 +6476,7 @@ log__open_auditfaillogfile(int logfile_state, int locked)
     char tbuf[TBUFSIZE];
     struct logfileinfo *logp;
     char buffer[BUFSIZ];
-    int rc;
+    int rc = 0;
 
     if (!locked)
         LOG_AUDITFAIL_LOCK_WRITE();
diff --git a/ldap/servers/slapd/modify.c b/ldap/servers/slapd/modify.c
index 58b606b2e..b0ec1c02c 100644
--- a/ldap/servers/slapd/modify.c
+++ b/ldap/servers/slapd/modify.c
@@ -598,6 +598,7 @@ modify_internal_pb(Slapi_PBlock *pb)
             if (pw_change == -1) {
                 /* The internal result code will already have been set by op_shared_allow_pw_change() */
                 ldap_mods_free(normalized_mods, 1);
+                slapi_ch_free_string(&old_pw);
                 return 0;
             }
         }
@@ -621,6 +622,7 @@ modify_internal_pb(Slapi_PBlock *pb)
     /* perform modify operation */
     slapi_td_internal_op_start();
     op_shared_modify(pb, pw_change, old_pw);
+    slapi_ch_free_string(&old_pw);
     slapi_td_internal_op_finish();
 
     /* free the normalized_mods don't forget to add this*/
@@ -1196,6 +1198,7 @@ op_shared_allow_pw_change(Slapi_PBlock *pb, LDAPMod *mod, char **old_pw, Slapi_M
     char *proxystr = NULL;
     char *errtext = NULL;
     int32_t needpw = 0;
+    bool free_bogus_entry = false;
 
     slapi_pblock_get(pb, SLAPI_IS_REPLICATED_OPERATION, &repl_op);
     if (repl_op) {
@@ -1255,6 +1258,7 @@ op_shared_allow_pw_change(Slapi_PBlock *pb, LDAPMod *mod, char **old_pw, Slapi_M
             e = slapi_entry_alloc();
             slapi_entry_init(e, NULL, NULL);
             slapi_sdn_set_dn_byref(slapi_entry_get_sdn(e), dn);
+            free_bogus_entry = true;
         }
 
         /* Set the backend in the pblock.  The slapi_access_allowed function
@@ -1400,6 +1404,9 @@ op_shared_allow_pw_change(Slapi_PBlock *pb, LDAPMod *mod, char **old_pw, Slapi_M
     valuearray_free(&values);
 
 done:
+    if (free_bogus_entry) {
+        slapi_entry_free(e);
+    }
     slapi_sdn_done(&sdn);
     slapi_ch_free_string(&proxydn);
     slapi_ch_free_string(&proxystr);
diff --git a/ldap/servers/slapd/operation.c b/ldap/servers/slapd/operation.c
index 97832db41..f70c5e035 100644
--- a/ldap/servers/slapd/operation.c
+++ b/ldap/servers/slapd/operation.c
@@ -223,6 +223,7 @@ operation_done(Slapi_Operation **op, Connection *conn)
             (*op)->o_results.result_controls = NULL;
         }
         slapi_ch_free_string(&(*op)->o_results.result_matched);
+        slapi_ch_free_string(&(*op)->o_results.result_text);
         int options = 0;
         /* save the old options */
         if ((*op)->o_ber) {
diff --git a/ldap/servers/slapd/pblock.c b/ldap/servers/slapd/pblock.c
index ec820ec70..3801326c4 100644
--- a/ldap/servers/slapd/pblock.c
+++ b/ldap/servers/slapd/pblock.c
@@ -4005,6 +4005,7 @@ slapi_pblock_set(Slapi_PBlock *pblock, int arg, void *value)
             pblock->pb_op->o_results.result_matched = (char *)value;
         break;
     case SLAPI_RESULT_TEXT:
+        /* value will get freed in operation_done(), so it should be allocated */
         if (pblock->pb_op != NULL)
             pblock->pb_op->o_results.result_text = (char *)value;
         break;
diff --git a/ldap/servers/slapd/plugin.c b/ldap/servers/slapd/plugin.c
index ba4a194c1..1528d5593 100644
--- a/ldap/servers/slapd/plugin.c
+++ b/ldap/servers/slapd/plugin.c
@@ -708,11 +708,6 @@ slapi_send_ldap_result_from_pb(Slapi_PBlock *pb)
     if (NULL != fn) {
         (*fn)(pb, err, matched, text, 0, NULL);
     }
-
-    slapi_pblock_set(pb, SLAPI_RESULT_TEXT, NULL);
-    slapi_pblock_set(pb, SLAPI_RESULT_MATCHED, NULL);
-    slapi_ch_free((void **)&matched);
-    slapi_ch_free((void **)&text);
 }
 
 void
diff --git a/ldap/servers/slapd/slapi-memberof.c b/ldap/servers/slapd/slapi-memberof.c
index 0b525896a..71df0f463 100644
--- a/ldap/servers/slapd/slapi-memberof.c
+++ b/ldap/servers/slapd/slapi-memberof.c
@@ -91,7 +91,7 @@ static int  sm_memberof_get_groups_callback(Slapi_Entry *e, void *callback_data)
 static void sm_report_error_msg(Slapi_MemberOfConfig *config, char* msg);
 static int  sm_entry_get_groups(Slapi_MemberOfConfig *config, Slapi_DN *member_sdn,
                                 Slapi_ValueSet *groupvals, Slapi_ValueSet *nsuniqueidvals);
-static PRBool sm_compare_memberof_config(const char *memberof_attr, char **groupattrs, PRBool all_backends, 
+static PRBool sm_compare_memberof_config(const char *memberof_attr, char **groupattrs, PRBool all_backends,
                                          PRBool skip_nested, Slapi_DN **include_scope, Slapi_DN **exclude_scope, PRBool enabled_only);
 static void sm_add_ancestors_cbdata(sm_memberof_cached_value *ancestors, void *callback_data);
 static int  sm_memberof_call_foreach_dn(Slapi_PBlock *pb __attribute__((unused)), Slapi_DN *sdn, Slapi_MemberOfConfig *config, char **types,
@@ -746,7 +746,6 @@ sm_entry_get_groups(Slapi_MemberOfConfig *config, Slapi_DN *member_sdn, Slapi_Va
                           "sm_entry_get_groups - Failed to retrieve target entry %s: %d\n",
                           slapi_sdn_get_ndn(group_sdn), rc);
             slapi_sdn_free(&group_sdn);
-            slapi_ch_array_free(groups_dn);
             rc = -1;
             goto common;
         }
@@ -767,7 +766,6 @@ sm_entry_get_groups(Slapi_MemberOfConfig *config, Slapi_DN *member_sdn, Slapi_Va
         sval = slapi_value_new_string(slapi_sdn_get_ndn(group_sdn));
         slapi_valueset_add_value_ext(groupvals, sval, SLAPI_VALUE_FLAG_PASSIN);
 
-
         slapi_sdn_free(&group_sdn);
         slapi_search_get_entry_done(&group_pb);
     }
diff --git a/ldap/servers/snmp/main.c b/ldap/servers/snmp/main.c
index 75dc3e76b..5250a2dbc 100644
--- a/ldap/servers/snmp/main.c
+++ b/ldap/servers/snmp/main.c
@@ -451,11 +451,10 @@ load_config(char *conf_path)
                             got_port = 1;
                         } else if (strcmp(attr, "nsslapd-rundir") == 0) {
                             /* 8 =  "/" + ".stats" + \0 */
-                            serv_p->stats_file = malloc(vlen + (instancename ? strlen(instancename) : 0) + 8);
+                            serv_p->stats_file = calloc(1, vlen + (instancename ? strlen(instancename) : 0) + 8);
                             if (serv_p->stats_file && instancename) {
-                                snprintf(serv_p->stats_file, vlen + strlen(instancename) + 8,
+                                snprintf(serv_p->stats_file, vlen + strlen(instancename) + 7,
                                          "%s/%s.stats", val, instancename);
-                                serv_p->stats_file[(vlen + strlen(instancename) + 7)] = (char)0;
                             } else {
                                 printf("ldap-agent: malloc error processing config file\n");
                                 free(entry);
@@ -466,6 +465,7 @@ load_config(char *conf_path)
                             }
                             got_rundir = 1;
                         }
+
                         free(attr);
                         free(val);
                         attr = NULL;
-- 
2.49.0

